p2p/pex/addrbook.go

```diff
@@ -306,13 +306,23 @@ func (a *addrBook) PickAddress(biasTowardsNewAddrs int) *p2p.NetAddress {
 			bucket = a.bucketsNew[a.rand.Intn(len(a.bucketsNew))]
 		}
 	}
-	// pick a random index and loop over the map to return that index
-	randIndex := a.rand.Intn(len(bucket))
+
+	// Create a slice of known addresses from the bucket
+	addresses := make([]*knownAddress, 0, len(bucket))
 	for _, ka := range bucket {
-		if randIndex == 0 {
+		addresses = append(addresses, ka)
+	}
+
+	// Shuffle the addresses
+	a.rand.Shuffle(len(addresses), func(i, j int) {
+		addresses[i], addresses[j] = addresses[j], addresses[i]
+	})
+
+	// Iterate through the shuffled addresses and apply the filter until we find a suitable address
+	for _, ka := range addresses {
+		if filter == nil || filter(ka) {
```

Comment:
Instead of picking a peer from a random index, we shuffle the bucket and return the first peer that satisfies the filter.

Reply:
Before what we would do is, pick a random index and stick with it, even if it doesn't satisfy filter requirements. By shuffling, we can start at the top of the list and work our way down to the first address that satisfies our filter. It essentially keeps the randomness while allow our selection pool to be much bigger. 

I don't think we need to shuffle, it just felt like an easier way to keep the randomness while keeping the pool of possibilities high. Happy to switch it out for something else if it makes sense!

Reply:
(in fact if I had to bet, I would say shuffling is less performant. Can you think of a better way to choose a random index, but keep choosing more and more random indexes until we pick a good candidate? I could think of a way if bucket wasn't a map and was instead an array. I guess I could make it an array and do many choices of random index, lmk what you think)

Reply:
Why not pick a random element and then check if it satisfies `filter`:

* if yes, then proceed with this element
* if not, pick another random element and check again

Reply:
Shuffle has the cost of a lock too, because of the `internal/rand` implementation.

By the way, this code also imports `math/rand`, so someone is not using our own implementation.

Reply:
The solution of taking random indexes in sequence is good if the probability of a random choice is good enough. If not, you are pretty likely to pick that index again. If you reject a lot of random choices, waiting for the random generator to pick a still not picked one is pretty low...

Reply:
But, again, why not to use the other method `GetSelection(bias)`. It does the shuffling, it returns a list of address with no repetitions, and you only need to call it once from the PEX reactor.

Reply:
I did not know the other method existed, am looking into using this now!

Reply:
I both removed the separate reserve map in favor of a larger toDial, as well as used GetSelection here https://github.com/cometbft/cometbft/pull/3371/commits/6cf054c314136ca0088bca05ac266bb85b8ee1cf

Still need to test it out, but seems like I definitely don't need to break the API to filter, thanks for the recommendation :)